optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsJoinBiConstraintStream.java

```diff
@@ -0,0 +1,142 @@
+        this.rightParentStream = otherStream;
+        this.biJoiner = (AbstractBiJoiner<A, B>) biJoiner;
+        this.rightPattern = otherStream.getPattern().expr(getLeftVariableDeclaration(), (b, a) -> matches(a, b));
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, A> getLeftParentStream() {
+        return leftParentStream;
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, B> getRightParentStream() {
+        return rightParentStream;
+    }
+
+    @Override
+    public Declaration<A> getLeftVariableDeclaration() {
+        return leftParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<A> getLeftPattern() {
+        return leftParentStream.getPattern();
+    }
+
+    @Override
+    public Declaration<B> getRightVariableDeclaration() {
+        return rightParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<B> getRightPattern() {
+        return rightPattern;
+    }
+
+    private static boolean matches(Object left, Object right, JoinerType type) {
+        switch (type) {
+            case EQUAL:
+                return left.equals(right);
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+            case GREATER_THAN:
+            case GREATER_THAN_OR_EQUAL:
+                return matchesComparable((Comparable)left, right, type);
+            case CONTAINING:
+                return ((Collection) left).contains(right);
+            case DISJOINT:
+            case INTERSECTING:
+                return matchesCollection((Collection)left, (Collection) right, type);
+            default:
+                throw new IllegalStateException("Unsupported joiner type (" + type + ").");
+        }
+    }
+
+    private static boolean matchesComparable(Comparable left, Object right, JoinerType type) {
+        int comparison = left.compareTo(right);
+        switch (type) {
+            case LESS_THAN:
+                return comparison < 0;
+            case LESS_THAN_OR_EQUAL:
+                return comparison <= 0;
+            case GREATER_THAN:
+                return comparison > 0;
+            case GREATER_THAN_OR_EQUAL:
+                return comparison >= 0;
+            default:
+                throw new IllegalStateException("Unsupported joiner type (" + type + ").");
+        }
+    }
+
+    private static boolean matchesCollection(Collection left, Collection right, JoinerType type) {
+        switch (type) {
+            case DISJOINT:
+                return left.stream().noneMatch(right::contains) && right.stream().noneMatch(left::contains);
+            case INTERSECTING:
+                return left.stream().anyMatch(right::contains) || right.stream().anyMatch(left::contains);
+            default:
+                throw new IllegalStateException("Unsupported joiner type (" + type + ").");
+        }
+    }
+
+    private boolean matches(A left, B right) {
```

Comment:
This is is not as performant as it could be.
It could use inlining:
Based on the joinerType(s), we can build a different lambda, to bring the switch(joinerType) cost to construction time.
Doesn't have to happen in this PR, but let's add a TODO in the constructor (at my other comment)

Reply:
I probably don't understand what you mean. Do you have an example of code somewhere, where you do that already?

Reply:
On second thought, I get it now. Although to fix this will make the code much less readable.

Reply:
On secound thoght, I get it now. Although fixing it will make the code much less readable.